pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ConsistentHashingStickyKeyConsumerSelector.java

```diff
@@ -44,15 +46,105 @@ public class ConsistentHashingStickyKeyConsumerSelector implements StickyKeyCons
 
     // Consistent-Hash ring
-    private final NavigableMap<Integer, List<Consumer>> hashRing;
+    private final NavigableMap<Integer, HashRingEntry> hashRing;
+    // used for distributing consumer instance selections evenly in the hash ring when there
+    // are multiple instances of consumer with the same consumer name or when there are hash collisions
+    private final Map<Consumer, MutableInt> consumerSelectionCounters;
 
     private final int numberOfPoints;
 
     public ConsistentHashingStickyKeyConsumerSelector(int numberOfPoints) {
         this.hashRing = new TreeMap<>();
+        this.consumerSelectionCounters = new WeakHashMap<>();
         this.numberOfPoints = numberOfPoints;
     }
 
+    /**
+     * This class is used to store the consumers and the selected consumer for a hash value in the hash ring.
+     * This attempts to distribute the consumers evenly in the hash ring for consumers with the same
+     * consumer name and priority level. These entries collide in the hash ring.
+     * The selected consumer is the consumer that is selected to serve the hash value.
+     * It is not changed unless a consumer is removed or a colliding consumer with higher priority or
+     * lower selection count is added.
+     */
+    private static class HashRingEntry {
+        // This class is used to store the consumer added to the hash ring
+        // sorting will be by priority, consumer name and active "selection" count of the consumer instance
+        // so that consumers get evenly distributed
+        record ConsumerEntry(Consumer consumer, MutableInt consumerSelectionCounter)
+                implements Comparable<ConsumerEntry> {
+            private static final Comparator<ConsumerEntry> CONSUMER_ENTRY_COMPARATOR =
+                    Comparator.<ConsumerEntry, Integer>
+                            // priority level is the primary sorting key
+                                    comparing(entry -> entry.consumer().getPriorityLevel()).reversed()
+                            // consumer name is the secondary sorting key
+                            .thenComparing(entry -> entry.consumer().consumerName())
+                            // then prefer the consumer instance with lowest selection count
+                            // so that consumers get evenly distributed
+                            .thenComparing(ConsumerEntry::consumerSelectionCounter);
+
+            @Override
+            public int compareTo(ConsumerEntry o) {
+                    return CONSUMER_ENTRY_COMPARATOR.compare(this, o);
+            }
+        }
+
+        private final List<ConsumerEntry> consumers;
+        ConsumerEntry selectedConsumerEntry;
+
+        public HashRingEntry() {
+            this.consumers = new ArrayList<>();
+        }
+
+        public Consumer getSelectedConsumer() {
+            return selectedConsumerEntry != null ? selectedConsumerEntry.consumer() : null;
```

Comment:
> This PR breaks the original behavior( the original behavior contains an issue ), which may cause a consumer not to receive any messages if users register consumers with the same name.

The hash ring will contain 100 entries and for each entry, the selected consumer will be evenly distributed. That's how it's already covered. Most of the complexity in this PR about addressing that issue that all registered consumers with the same name would get evenly distributed. 
This applies when there are less than 100 consumers. This can be configured by increasing `subscriptionKeySharedConsistentHashingReplicaPoints` if that's not sufficient. However, users should use unique consumer names if there are more than 100 consumers.

Reply:
I think you have not answered my questions, such as follows

> The issue's occurring condition is that multiple consumers have the same hash( it always occurs when the consumer names are the same ), right?

> This PR breaks the original behavior( the original behavior contains an issue ), which may cause a consumer not to receive any messages if users register consumers with the same name.

Reply:
> > This PR breaks the original behavior( the original behavior contains an issue ), which may cause a consumer not to receive any messages if users register consumers with the same name.

> No. I explained in the previous comment.

- The consumers that were calculated with the same `hash` will be pushed to the same `HashRingEntry`
- The selector will always get the same `HashRingEntry` for a certain `key`
- The `HashRingEntry` selected will always return the consumer who has the highest priority.
- The consumers who have a lower priority will never be selected

Reply:
> - The consumers that were calculated with the same `hash` will be pushed to the same `HashRingEntry`
> - The selector will always get the same `HashRingEntry` for a certain `key`

That's the correct behavior and the intent of this PR. The previous behavior was not correct at all since it caused the bug.

There will be multiple hash ring entries (100 by default) and the colliding consumers are evenly selected across all of them. Did you already consider that aspect?


> - The `HashRingEntry` selected will always return the consumer who has the highest priority.
> - The consumers who have a lower priority will never be selected

It's true. The consumer priority was ignored before, but that has been a bug. Bugs should be fixed.

Reply:
>> - The consumers that were calculated with the same `hash` will be pushed to the same `HashRingEntry`
>> - The selector will always get the same `HashRingEntry` for a certain `key`

> There will be multiple hash ring entries (100 by default) and the colliding consumers are evenly selected across all of them. Did you already consider that aspect?

Yes, but the multiple consumers who have the same name will be assigned the same 100 `HashRingEntry`s, right?

If yes, the consumers who have a lower priority will never be selected, they will receive nothing.

Reply:
> Yes, but the multiple consumers who have the same name will be assigned the same 100 `HashRingEntry`s, right?

No. Most of the complexity in this PR is handle the case that multiple consumers with the same name will be evenly distributed as the assigned consumers for the 100 hash ring points.
Please take a look at the diff and the code comments for example for the `consumerSelectionCounters` field and how that gets used. The counter for each consumer keeps the state of the number of hash ring entries where a specific consumer is "selected" (assigned as the consumer). This counter is part of the ordering criterias for the ConsumerEntry. That's how it's determined that for consumers with the same priority and name, the lowest selection count consumer gets selected. Is the current code understandable or do you think that more comments should be added to explain this "magic"?

> If yes, the consumers who have a lower priority will never be selected, they will receive nothing. You broke the bahavior

Why would they set the priority if they don't want it to be used? I wouldn't call this in breaking behavior. We can also ask @merlimat for an opinion.

> I think consumer.priority is designed for Failover mode, so Key_shared discarded it.

It's not only for Failover mode. The javadoc explains that it's support for Shared mode too: https://github.com/apache/pulsar/blob/cefa72c7d18ab05d413e8737672f746629874f59/pulsar-client-api/src/main/java/org/apache/pulsar/client/api/ConsumerBuilder.java#L508-L553 . It's a matter of defining what the behavior should be for Key_Shared subscriptions.

Reply:
>> Yes, but the multiple consumers who have the same name will be assigned the same 100 `HashRingEntry`s, right?

> No. Most of the complexity in this PR is handle the case that multiple consumers with the same name will be evenly distributed as the assigned consumers for the 100 hash ring points.
Please take a look at the diff and the code comments for example for the `consumerSelectionCounters` field and how that gets used. The counter for each consumer keeps the state of the number of hash ring entries where a specific consumer is "selected" (assigned as the consumer).

Yes, I noticed `consumerSelectionCounters` before, but the values of this map are always `0`, which is meanless

Reply:
> Yes, I noticed `consumerSelectionCounters` before, but the values of this map are always `0`, which is meanless

I don't see that happening. You could modify ConsistentHashingStickyKeyConsumerSelectorTest#shouldRemoveConsumersFromConsumerKeyHashRanges for example and make it use 100 points and a duplicate name. The counters are 1 in most cases which is a sign that this "magic" really works. There are some other Key_Shared tests that would fail if this didn't work.
<img width="1244" alt="image" src="https://github.com/user-attachments/assets/65dbb170-cbf6-45cf-973c-9b6bf584561a">
Do you see it now?

Reply:
> Do you see it now?

Sure, I found it now, it was changed by `beforeChangingSelectedConsumerEntry` and `afterChangingSelectedConsumerEntry`.  

Since you used it to mark how many `points` it owned, your PR has another issue; I wrote a test to reproduce it.

```java
    @Test
    public void testConsumerReconnects() {
        final ConsistentHashingStickyKeyConsumerSelector selector = new ConsistentHashingStickyKeyConsumerSelector(100);
        final String consumerName = "consumer";
        final int numOfInitialConsumers = 50;
        final int validationPointCount = 200;
        final List<Integer> pointsToTest = pointsToTest(validationPointCount);
        List<Consumer> consumers = new ArrayList<>();
        for (int i = 0; i < numOfInitialConsumers; i++) {
            final Consumer consumer = createMockConsumer(consumerName, "index " + i, i);
            consumers.add(consumer);
            selector.addConsumer(consumer);
        }

        // Mark original results.
        List<Consumer> selectedConsumersBeforeRemove = new ArrayList<>();
        for (int i = 0; i < validationPointCount; i++) {
            int point = pointsToTest.get(i);
            selectedConsumersBeforeRemove.add(selector.select(point));
        }

        // Let consumers reconnect
        // Verify: the same range selected as "selectedConsumersBeforeRemove"
        for (Consumer c : consumers) {
            selector.removeConsumer(c);
        }
        for (int i = numOfInitialConsumers -1; i > -1 ; i--) {
            // Consumers reconnect.
            selector.addConsumer(consumers.get(i));
            for (int j = 0; j < validationPointCount; j++) {
                int point = pointsToTest.get(j);
                Consumer selected = selector.select(point);
                Consumer expected = selectedConsumersBeforeRemove.get(j);
                assertThat(selected.consumerId()).as("validationPoint %d, hash %d", j, point).isEqualTo(expected.consumerId());
            }
        }
    }

    private List<Integer> pointsToTest(int validationPointCount) {
        List<Integer> res = new ArrayList<>();
        int hashRangeSize = Integer.MAX_VALUE;
        final int increment = hashRangeSize / (validationPointCount + 1);
        for (int i = 0; i < validationPointCount; i++) {
            res.add(i * increment);
        }
        return res;
    }
```

Reply:
> Since you used it to mark how many `points` it owned, your PR has another issue; I wrote a test to reproduce it.

@poorbarcode Thanks, this is a good way to handle the details. I'll check what could be supported in the solution. I provided a longer update in a separate comment https://github.com/apache/pulsar/pull/23327#issuecomment-2376288786. Without an additional key, it's not possible to solve this problem. The challenge is that the client IP address is mostly useless in many k8s deployments which don't forward the original client IP address.

I've taken an approach where I'm trying to add an additional key based on the arrival index of a specific consumer name. If the index is unused, it can be taken. This way a consumer that leaves and joins back would be possible to get the same assignment. However the code doesn't yet work properly.

Reply:
@poorbarcode regarding the unit test `testConsumerReconnects`, does that pass with current master branch?

Reply:
> @poorbarcode regarding the unit test testConsumerReconnects, does that pass with current master branch?

It can not pass, but what I want to say is you considered Pulsar should keep the key range for a consumer absolutely consistent, but you just fixed half of the issue. If you agree that you just want to solve half of the issue, I could provide a suggestion that is simpler.

Reply:
> It can not pass, but what I want to say is you considered Pulsar should keep the key range for a consumer absolutely consistent, but you just fixed half of the issue. If you agree that you just want to solve half of the issue, I want to advise a new solution that is simpler than the current one.

This issue is primarily fixing the out-of-order issue which is a critical issue, #23315. It is caused with the fact that the current implementation will swap 2 existing consumers when other consumers connect or disconnect. 
Let's first make progress on addressing that issue in this PR. Ok?

Reply:
> I tried the [test above](https://github.com/apache/pulsar/pull/23327#discussion_r1776320513) again, it can not pass even if I pulled the newest commits

Yes that's true, I haven't had a chance to dig in the details yet.

Reply:
@poorbarcode The test is fixed in 80d64514fa35daa4d6e507c92b43a3a3313835bf . The problem in your test was that it was assuming that the assignments would be the same when there were different consumer counts. The current implementation will ensure that when consumers leave and join in the same order, they will get assigned with the same hash range assignments. This is a reasonable "best effort" approach when the application is assigning the same consumer name. As we have discussed, that is not a recommended approach in the first place for Key_Shared subscriptions. This PR is for fixing #23315 and #23321 with a reasonable approach.